Adding s3 encryption

jaustinpage opened the pull request:
Add optional s3 encryption

jaustinpage committed:
Adding s3 encryption

jaustinpage commented:
output works:

```shell
$ terraform output -module='remote_state'
dynamodb_lock_stream_arn = [
    arn:aws:dynamodb:us-west-2:201770704483:table/terraform-lock/stream/2018-11-12T21:27:17.753
]
dynamodb_lock_stream_label = [
    2018-11-12T21:27:17.753
]
dynamodb_lock_table_arn = [
    arn:aws:dynamodb:us-west-2:201770704483:table/terraform-lock
]
dynamodb_lock_table_name = [
    terraform-lock
]
s3_backend_bucket_name = pwpwn-v2-tfstate
```

jaustinpage commented:
also, when https://github.com/hashicorp/terraform/issues/14037 lands, we should adjust this to avoid the `count="{$var.turns_on_off}`  hacks.

jaustinpage commented:
addresses: https://github.com/samstav/terraform-aws-backend/issues/8

samstav commented:
@jaustinpage thank you! This is going to be a great addition.

I see that this defines a separate/different resource in terraform if encryption is enabled. I can't think of a way around this, and I suspect you already considered it.

I would love to get this feature added, and for any existing projects using this module to be able to switch to using encryption. I'll try it with this branch, but I assume it won't work, at least not easily-- unless users `terraform import aws_s3_bucket.tf_backend_bucket_encrypted <existing_bucket>` ?

jaustinpage commented:
@samstav yes, your points are well received. Until we see examples of https://github.com/hashicorp/terraform/issues/14037 being fixed, i will implement this so that encryption is enabled by default.

This will have a net effect of costing operators $1 per month (unless they run terraform more than ~10k times in a month), but, this seems like a reasonable tradeoff compared to having the s3 bucket with the terraform.

I will implement this with encryption set to default, will create a new issue to track implementing the optional behavior when terraform implements something useful in a clean way.

I will also expose the ``kms_master_key_id``.

might take me a few days to get back to you on this.

samstav commented:
@jaustinpage Brilliant!

stavxyz committed:
remove encryption enabled variable (always on now)

stavxyz committed:
add variables for kms key and sse algo

stavxyz committed:
refactor bucket policy with aws_iam_policy_document

stavxyz committed:
consolidate to one encrypted bucket resource (using new vars)

samstav commented:
https://github.com/jaustinpage/terraform-aws-backend/pull/1

jaustinpage committed:
Merge pull request #1 from samstav/samstav/s3-encryption

updates to the s3 encryption work

stavxyz commented on main.tf:
```diff
@@ -75,11 +75,69 @@ resource "aws_s3_bucket" "tf_backend_bucket" {
+      values = [
+        "AES256"
+      ]
+    }
```
Outstanding questions... will this bucket policy block requests if the SSE Algorithm is set to `AES256`? Need to check that and make sure that either algorithm will work with this policy, OR.... is this crazy?

```suggestion
    condition {
      test = "StringEquals"
      variable = "s3:x-amz-server-side-encryption"
      values = [
        "${var.sse_algorithm}"
      ]
    }
```

jpbochi replied:
I think you can deny unencrypted storage by doing this:
```
      "Condition": {
        "Null": {
          "s3:x-amz-server-side-encryption": "true"
        }
      }
```

source: https://docs.aws.amazon.com/AmazonS3/latest/dev/UsingServerSideEncryption.html

jaustinpage replied:
OK, here is where i landed:
1) it is stronger to "reject invalid cases". so the existing "deny if not equal to" style rule will persist
2) i did miss getting the var in here, i have corrected this.

stavxyz commented on variables.tf:
```diff
@@ -25,3 +25,14 @@ variable "lock_table_read_capacity" {
+
+variable "kms_key_id" {
+  # Default to absent/blank to use the default aws/s3 aws kms master key
+  default = ""
```
Need to make sure that this *actually* works. There is a chance that this will actually try to use a literal blank string as the key id, which will likely fail on the bucket creation/modification. Even if it doesn't fail, we need to make sure that this default setting `""` results in the use of the `aws/s3 kms master key`.

jaustinpage replied:
I have manually verified that this uses the master kms key.

jpbochi commented on README.md:
```diff
@@ -46,6 +46,7 @@ module "backend" {
   backend_bucket = "terraform-state-bucket"
   # using options, e.g. if you dont want a dynamodb lock table, uncomment this:
   # dynamodb_lock_table_enabled = false
+  # s3_state_encryption_enabled = false
```
did you intend to add this variable?

jaustinpage replied:
good eye

jaustinpage committed:
Fixes for the bucket encryption, update docs

jaustinpage committed:
Fixing a few typeos

jaustinpage committed:
update the way principals are specified

samstav commented:
This is looking really good. I'll want to test this out manually now

https://github.com/samstav/terraform-aws-backend/pull/7#discussion_r244792765

jaustinpage committed:
Add the kms key output, docs

updated the output to include the master kms key for the account if no kms key was specified. also, updated the docs to specify that you should specify the kms key in the s3 backend config

stavxyz commented on README.md:
```diff
@@ -91,7 +91,7 @@ terraform init -backend=false
 # Step 4: Apply the infrastructure plan
 terraform apply backend.plan
-# Step 5: Only after applying (building) the backend resources, write our terraform config
+# Step 5: Only after applying (building) the backend resources, write our terraform config. use the terraform apply output to find your kms key id, and substitude it into the command below.
```
```suggestion
# Step 5: Only after applying (building) the backend resources, write our terraform config. Use the `terraform apply` output to find your kms key id, and substitute it into the command below.
```

jaustinpage replied:
removed this in favor of simplified interface

stavxyz commented on README.md:
```diff
@@ -100,6 +100,8 @@ echo 'terraform { backend "s3" {} }' > conf.tf
     -backend-config="bucket=terraform-state-bucket" \
     -backend-config="key=states/terraform.tfstate" \
+    -backend-config="encrypt=1" \
+    -backend-config="kms_key_id=YOUR_KMS_KEY_ID" \
```
Is this always required? It uses the default key if not specified, right?

jaustinpage replied:
i have simplified the examples to reflect the improved defaults.

stavxyz commented on outputs.tf:
```diff
@@ -29,3 +30,7 @@ output "dynamodb_lock_stream_arn" {
 }
+
+output "s3_kms_key_id" {
+  value = "${var.kms_key_id == "" ? var.kms_key_id : data.aws_kms_key.master.arn}"
```
Is the arn different than the id? My understanding of that data source is that it will accept an alias OR an ID, which is a nice convenience provided by terraform (if I understand correctly).

jaustinpage replied:
i removed this in favor of a simplified kms key specification interface

stavxyz commented on main.tf:
```diff
@@ -96,3 +156,6 @@ resource "aws_s3_bucket" "tf_backend_logs_bucket" {
 }
 
+data "aws_kms_key" "master" {
+  key_id = "alias/aws/s3"
```
I wonder if you could/should do this?

```
data "aws_kms_key" "master" {
  key_id = "${var.kms_key_id == "" ? "alias/aws/s3" : var.kms_key_id}"
}
```
This would demand its existence on the tf run though.........

But you could modify your output variable to a static reference to this data output.

I am not sure if there are any other implications here. Let me know what your thoughts are on this.

jaustinpage replied:
this seems cleaner, i like it

jaustinpage replied:
i dont think we need to worry about having this as an output.

If a user specifies a kms key, then they will also have to specify the same key in their conf.tf. 

if they dont specify a kms key, then they will not have to specify the key in their conf.tf.

samstav commented:
Unless the kms key variable is required for basic functionality, I think it would be good to leave the examples/docs without its use and simply document the availability of the option/variable.

jaustinpage commented:
> Unless the kms key variable is required for basic functionality, I think it would be good to leave the examples/docs without its use and simply document the availability of the option/variable.

so, we have a choice. If we set https://github.com/samstav/terraform-aws-backend/pull/7/files#diff-7a370d8342e7203b805911c92454f0f4R127 to `AES256` by default, then we do not have to set the kms key variable, as with `encrypt = true` in the conf.tf file, aes256 will be satisfied.

however, with the current default of `aws:kms`, both `encrypt = true` and the kms key id are required in the conf.tf file in order for s3 to not reject the s3 bucket update.

Note, that this is independent of https://github.com/samstav/terraform-aws-backend/pull/7/files#diff-7a370d8342e7203b805911c92454f0f4R82 ; which will allow either method (or none), but will turn unencrypted updates into encrypted updates. 

to me, it basically feels like we either have to split the sse_algorithm variable, to allow for splits, OR we have to manually specify the kms key in the conf.tf, even if we are using the master one. Putting the kms key (with assistance in the output vars), even though it is more cumbersome, allows for a single set of instructions about how to use the module, while also providing the ease of specifying your own kms key if you choose.

One other approach would be to, if the sse_algorithm var is not specified, have it default to `aws:kms` in the https://github.com/samstav/terraform-aws-backend/pull/7/files#diff-7a370d8342e7203b805911c92454f0f4R82 , and then default to `aes256` for https://github.com/samstav/terraform-aws-backend/pull/7/files#diff-7a370d8342e7203b805911c92454f0f4R127, but this seemed less transparent to an end user, if more convenient. 

I dont have strong feelings either way...

jaustinpage committed:
Improved encryption usability

Simplified encryption of s3 bucket to default to s3 encryption method, and automatically switch to kms encryption if a kms key is specified. Updated the docs to reflect the simpler interface.

jaustinpage committed:
Clean up readme to reflect simplified interface

jaustinpage commented:
I have manually verified the "Commands are the fun part" https://github.com/jaustinpage/terraform-aws-backend#commands-are-the-fun-part section, steps 1-5, but using a terraform conf.tf file, and modifying step 3 to be `terraform plan -out=backend.plan` ~~-target=module.backend~~. I did this both specifying a kms key (and specifying the full arn for the `kms_key_id` in the conf.tf file), and not specifying a kms key (and not specifying a `kms_key_id` in the conf.tf file). Both times, I was able to successfully finish with a `terraform init -reconfigure`, saying yes to updating the state, and getting a successful terraform exit.

TLDR: "works on my machine"

Here is the `conf.tf` that i was using with kms key specified:
```hcl
terraform {
  backend "s3" {
    region         = "us-west-2"
    bucket         = "pwpwn-v6-tfstate"
    key            = "states/terraform.tfstate"
    encrypt        = "true"
    dynamodb_table = "pwpwn-v6-tflock"
    kms_key_id     = "arn:aws:kms:us-west-2:000000000000:key/00000000-0000-0000-0000-000000000000"
  }
}
```

and here is the `conf.tf` that i was using when not specifying the kms key:

```hcl
terraform {
  backend "s3" {
    region         = "us-west-2"
    bucket         = "pwpwn-v6-tfstate"
    key            = "states/terraform.tfstate"
    encrypt        = "true"
    dynamodb_table = "pwpwn-v6-tflock"
  }
}
```

samstav commented:
I only just realized that the `kms_key_id` must be specified in the terraform block if it is used!

https://www.terraform.io/docs/backends/types/s3.html#kms_key_id

Now your story makes sense. Thank you!!

stavxyz commented on variables.tf:
```diff
@@ -25,3 +25,9 @@ variable "lock_table_read_capacity" {
 }
+
+variable "kms_key_id" {
+  # Default to absent/blank to use the default aws/s3 aws kms master key
```
Ok, so this comment is no longer true, I think.

If the user wants to use the default aws/s3 AWS KMS master key then they must explicitly set the `kms_key_id` value to `alias/aws/s3`. Is that true?

Assuming that is true, must they also explicitly set the `kms_key_id` value in their [terraform backend configuration block](https://www.terraform.io/docs/backends/types/s3.html#kms_key_id) to `alias/aws/s3`?

jaustinpage replied:
No, terraform will automatically use the default account master kms key for you. This comment holds true

jaustinpage commented:
anything i can help test at this point?
